### DISCUSSION TITLE
Joined inheritance with non-primary parent key is not working.

### DISCUSSION THREAD

**Discussion by greemo** [sqlalchemy/sqlalchemy#7357] [Usage Questions] [answered]
### Describe the bug

I have a situation where postal addresses are used for multiple purposes in the system, so I created a base class PostalAddress and subclasses for each use-case where they are used as part of the composition of another object. For example, I have an Organisation that has a postal address. This is a 1:1 relationship.

I want to ensure addresses don't exist without an owner, so the foreign key is declared on the address side, rather than the organisation side.

I want to look up an OrganisationAddress from the session based on the id of the organisation. I have tries a few ways: 
```
address = await session.get(OrganisationAddress, {"organisation_id": org.node_id})
```
Outcome: "sqlalchemy.exc.InvalidRequestError: Incorrect names of values in identifier to formulate primary key for query.get(); primary key attribute names are 'id'"
```
address = await session.get(OrganisationAddress, org.node_id)
```
Outcome: Wrong SQL emitted: ``` SELECT organisation_address.id AS organisation_address_id, postal_address.id AS postal_address_id, postal_address.supplement_a AS postal_address_supplement_a, organisation_address.organisation_id AS organisation_address_organisation_id 
FROM postal_address JOIN organisation_address ON postal_address.id = organisation_address.id 
WHERE postal_address.id = %s```

I want to use the session.get() method, because I want to leverage the caching of the object when available.

However the "organisation_id" attribute of OrganisationAddress is not being recognised as a primary key, even though it should be, as it is THE primary key, as confirmed by the DB.

### To Reproduce

```python
class Organisation(BaseModel):
    __tablename__ = 'organisation'
    node_id = sa.Column(sa.Integer, primary_key=True, nullable=False)
    address = sa.orm.relationship('OrganisationAddress', uselist=False)
    ...

class PostalAddress(BaseModel):
    __tablename__ = 'postal_address'
    id = sa.Column(sa.Integer, primary_key=True, nullable=False)
    ...

# The primary key for the derived class is not the foreign key to the parent class, but a foreign key to an owning class.
class OrganisationAddress(PostalAddress):
    __tablename__ = 'organisation_address'
    id = sa.Column(sa.ForeignKey("postal_address.id"), nullable=False, index=True)
    organisation_id = sa.Column(sa.ForeignKey("organisation.node_id", ondelete='CASCADE'), primary_key=True, nullable=False, doc="Organisation owning this address")
```

### Error

```
Incorrect names of values in identifier to formulate primary key for query.get(); primary key attribute names are 'id'
Traceback (most recent call last):
/orm/session.py", line 2811, in _get_impl
    primary_key_identity = list(
/orm/session.py", line 2812, in <genexpr>
    primary_key_identity[prop.key]
KeyError: 'id'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
/execute.py", line 625, in await_result
    return_type, field_nodes, info, path, await result
in resolve
    result = await result
  File "organisation.py", line 15, in resolve_address
    address = await session.get(OrganisationAddress, {"organisation_id": org.node_id})
/session.py", line 295, in get
    return await greenlet_spawn(
/util/_concurrency_py3k.py", line 123, in greenlet_spawn
    result = context.switch(*args, **kwargs)
/orm/session.py", line 2768, in get
    return self._get_impl(
/orm/session.py", line 2817, in _get_impl
    util.raise_(
/util/compat.py", line 207, in raise_
    raise exception
sqlalchemy.exc.InvalidRequestError: Incorrect names of values in identifier to formulate primary key for query.get(); primary key attribute names are 'id'


``` 

### Versions

- OS: linux
- Python: 3.8
- SQLAlchemy: 1.4.26
- Database: Postgres 12.9
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

### Additional context

Using Async Session

**Comment by zzzeek** [comment]
the primary key attribute for an OrganizationAddress model is called "id", because this is the PostalAddress.id attribute.  All objects that are instances of PostalAddress, which includes objects that are subclasses of it, are identified by PostalAddress.id first and foremost since this is joined table inheritance.  The OrganizationAddress.organization_id attribute being a primary key is fine but that's a persistence detail.   so for the dictionary form here I think you'd want to use `{"id": ident}`, as it says.    

> Outcome: Wrong SQL emitted: SELECT organisation_address.id AS organisation_address_id, postal_address.id AS postal_address_id, postal_address.supplement_a AS postal_address_supplement_a, organisation_address.organisation_id AS organisation_address_organisation_id FROM postal_address JOIN organisation_address ON postal_address.id = organisation_address.id WHERE postal_address.id = %s

that's the correct SQL.    PostalAddress.id is the object identifier for all objects that are an instance of PostalAddress.

if you want to query the OrganizationAddress table alone, you might be looking to use [concrete table inheritance](https://docs.sqlalchemy.org/en/14/orm/inheritance.html#concrete-table-inheritance)  if you dont want PostalAddress to be part of it and you'd like to get OrganizationAddress objects from that one table alone.

**Reply by greemo** [reply]
Thanks for the response Mike,

So even though I explicitly set the PK to be OrganisationAddress.organisation_id for the OrganisationAddress, sqlalchemy overrides this and sets OrganisationAddress.id as the primary key, because this is the foreign key into the parent class? And this is intentional?

If so, I think the user should be warned that the primary keys they defined are ignored. Better yet, don't let them define a primary key other than the foreign key to the parent table.

This burned a good few hours of my time.

I'll dive deeper into concrete table inheritance, thanks.

**Comment by zzzeek** [comment, marked as answer]
> Thanks for the response zzzeek,
> 
> So even though I explicitly set the PK to be OrganisationAddress.organisation_id for the OrganisationAddress, sqlalchemy overrides this and sets OrganisationAddress.id as the primary key, because this is the foreign key into the parent class? 

so, nothing was "overridden" here, your "organization_address" table still has the organization_id column, and it still composes the primary key for that table in the database schema.  the mapping just does not normally consider this to be part of the object identity space of PostalAddress objects.  Because all PostalAddress objects that are not OrganizationAddress objects won't have this column at all!  so, it's not really natural as an object identity key for joined inheritance.

> And this is intentional?

as above, joined inheritance is mapping a class hierarchy.   It creates a set of identity keys that can identify any object within the hierarchy as a whole, and every class that extends from PostalAddress using joined inheritance will have a row in the postal_address table, therefore postal_address.id is sufficient to form the object identifier for all objects in the hierarchy and it's also the only key that is present on all objects in the hierarchy.


> 
> If so, I think the user should be warned that the primary keys they defined are ignored. 

the primary key constraint you defined was not ignored, it's on your table and is active against rows populated for that table.  The ORM will use it when it emits UPDATE and DELETE statements for the table as well.   All tables need to have one or more columns defined that comprise the primary key, whether or not they are the base table in a joined inheritance hierarchy.

> Better yet, don't let them define a primary key other than the foreign key to the parent table.

this is a valid mapping and is actually how we used to document this originally before we figured out that having the subtable FK and PK be the same column was probably clearer.   this is in the eighth paragraph at https://docs.sqlalchemy.org/en/14/orm/inheritance.html#joined-table-inheritance:

> In a polymorphic setup, it is most common that the foreign key constraint is established on the same column or columns as the primary key itself, however this is not required; a column distinct from the primary key may also be made to refer to the parent via foreign key. The way that a JOIN is constructed from the base table to subclasses is also directly customizable, however this is rarely necessary.


> 
> This burned a good few hours of my time.

OK, sorry about that, I think consulting the docs is best here.   If you look at https://docs.sqlalchemy.org/en/14/orm/inheritance.html#joined-table-inheritance, while the PK behavior is kind of implied for the first few paragraphs, in order to make sure this concept is clear there is a separate box that discusses this directly (relevant emphasis is added):

> Joined inheritance primary keys
> One natural effect of the joined table inheritance configuration is that the identity of any mapped object can be determined entirely from rows in the base table alone. This has obvious advantages, so **SQLAlchemy always considers the primary key columns of a joined inheritance class to be those of the base table only**. In other words, **the id columns of both the engineer and manager tables are not used to locate Engineer or Manager objects - only the value in employee.id is considered.** engineer.id and manager.id are still of course critical to the proper operation of the pattern overall as they are used to locate the joined row, once the parent row has been determined within a statement.


> 
> I'll dive deeper into concrete table inheritance, thanks.

note that if you do want to still join to the postal_address table you could probably map OrganizationAddress using concrete, but map it to a join of the two tables using the pattern at https://docs.sqlalchemy.org/en/14/orm/nonstandard_mappings.html#mapping-a-class-against-multiple-tables .

**Reply by greemo** [reply]
Thank you very much for your detailed response Mike. I can now see why this is needed.

I'm sure you have spent enough time on this already, and I don't want to waste your time from implementing features and bugfixes, so thank you, feel free to drop this.

It is very possible I am the one with a hole in my thinking, but if anyone else sees this post, and hits the same issue, please comment, and we'll see if we can make this error message clearer to someone.

Basically I still find the error from ```address = await session.get(OrganisationAddress, {"organisation_id": org.node_id})``` confusing.
```
sqlalchemy.exc.InvalidRequestError: Incorrect names of values in identifier to formulate primary key for query.get(); primary key attribute names are 'id'
```
I am doing a session.get() on the OrganisationAddress table, not the PostalAddress table, so I expect the primary key to be used to be the one from the table I passed to the get() method. i.e. the OrganisationAddress table.

Maybe the way it is implemented gives it a good reason to behave that way, but from the outside looking in, this behaviour is not natural, nor does it match the documentation on session.get() https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.get

> Return an instance based on the given primary key identifier, or None if not found

Based on this, I would expect the session.get() to lookup based on the primary key of the table passed in, then if it is joined with a parent via inheritance, also load the parent columns into the returned object.

Maybe a comment in the session.get() docs to say the primary key has to be the primary key of the root table in joined table inheritance would explain things a little further, but the behaviour of the get() method itself would then remain conditional to the type of instance passed in, and as such still unclear (to me.. see note above about holes in my thinking :)

**Reply by zzzeek** [reply]
i think the solution is that the documentation and the exception message which uses the phrase "primary key identiifer" for ORM methods must be changed to use a different term, which we refer to as an "identity key".  This is an existing term in the ORM but should be made into a linked glossary term which can get into the difference between "identity key" and "primary key".     added #7360 for this.
